### ISSUE TITLE
Endpoints controller issues maaany lists

### ISSUE THREAD

**Issue by fgrzadkowski** [GoogleCloudPlatform/kubernetes#6518]
Endpoints controller issues many lists for services and pods. The very heavy ones are those for pods. For each service it issues one list with a selector every 10 s. However even if we are interested in a single pod, apiserver will parse jsons for all pods. According to my profiling this constitutes for more than 60% of CPU usage in density test causing very high latency for all other requests.

@wojtek-t @davidopp @lavalamp

**Comment by wojtek-t** [comment]
Filip - IIUC you were running "Density" e2e test as your underlying experiment. In this test there are no services at all - what services did you have so that it was generating so much load?

**Comment by fgrzadkowski** [comment]
From the apiserver logs this looked it was for default services.

-- sent from a mobile device
7 kwi 2015 18:32 "Wojciech Tyczynski" <notifications@github.com> napisał(a):

**Comment by lavalamp** [comment]
It should use the local caching mechanism that e.g. scheduler uses, then this wouldn't be an issue. (controller manager framework).

**Comment by bgrant0607** [comment]
cc @thockin

**Comment by thockin** [comment]
I'm looking at endpoint controller now - it's pretty dumb.   It doesn't
watch or anything.

**Comment by markturansky** [comment]
Of all the controllers that are simple sync loops, this one is the "slightly different" one.

A good refactoring for this controller would be to split the looping logic and processing logic.  Make a "syncHandler" like the other controllers.  This would make it easier to port to a plugin (see #6520).

**Comment by davidopp** [comment]
@lavalamp said he will look into this. It will be fixed by the controller framework.

**Comment by fgrzadkowski** [comment]
Assigning to @lavalamp who said he will look into this.

**Comment by fgrzadkowski** [comment]
@lavalamp Are you working on this? If not, I'm happy to look into this. It'd be great to fix this ASAP so that we can move forward with profiling large clusters.

**Comment by lavalamp** [comment]
It's on my list for today.

**Comment by lavalamp** [comment]
This fell off the end of my queue. I will take it up tomorrow.

**Comment by lavalamp** [comment]
OK I have this mostly working but still need to clean it up/fix tests/fix some other bugs I found.

**Comment by lavalamp** [comment]
@fgrzadkowski would love to know if this made any detectable difference. :)

**Comment by fgrzadkowski** [comment]
I'm going to run profiling today. Will update this issue once I have some data.

**Comment by fgrzadkowski** [comment]
I did some profiling. First of density test runs faster and more predictable, which is great! However I still see high CPU usage, but with a different profile. According to initial observations we spend majority of time in sending modifications via watch. I need to dig deeper to understand this better.

I believe we managed to fix first bottleneck (we still need to fix replication controller, which I worked around in my tests) but we just hit the next one :)
